### PR TITLE
feat(appset): Advanced Templating using templatePatch

### LINKED ISSUE #9177: Support for multiple sources in application set with generators
# Summary
I used argoCD application-set to  deploy my helm charts from one source and my k8s manifests from another source. But there is no support for multiple sources in application-set in `argoproj.io/v1alpha1` API  version. 

# Motivation
To deploy different application sourced from different points, like helm charts from some repo at bitnami and customize packages from my own private repo. 
# Proposal

Multiple sources must be support in applicationset so that we can deploy multiple application by getting deployment files from different sources.

### LINKED ISSUE #11164: Allow advanced templating in ApplicationSet
People want to be able to do advanced things in ApplicationSet templates.

One way to accomplish this is to provide a plain ol' string field for the template instead of an Application spec. But then you lose nice IDE support.

Another way, which I think would require fewer changes, is to allow templating _keys_ instead of just values in the Application spec.

Something like this:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
spec:
  goTemplate: true
  generators:
  - list:
      elements:
      - automated: "true"
      - automated: "false"
  template:
    spec:
      syncPolicy:
        "{{ ternary "automated" "turtles" (eq .automated "true") }}": {}
```

When `automated` is not `"true"`, the key nested under `syncPolicy` evaluates to `turtles`. Since that's not part of the Application spec, it'll be dropped when the Application is unmarshaled. 

Similar strategies could be applied to toggle, for example, fields under `source`. That would allow a git files generator to generate applications both for Helm and Kustomize.

**Changes required:**

First, we'd have to change the ApplicationSet CRD to preserve unknown fields. `"{{ ternary "automated" "turtles" (eq .automated "true") }}"` would otherwise be rejected as an invalid key (or, if validation is disabled, dropped).

Second, we'd have to change the ApplicationSet controller to handle the ApplicationSpec as an unstructured until the very last moment, when _all_ templating has been applied. If we unmarshal to an Application spec before then, we'll lose the templated key.

Inspiration for this strange idea comes from fixing this bug, when I realized we can template keys: https://github.com/argoproj/argo-cd/pull/11163

### PR DESCRIPTION
**Pull request by speedfl**
Signed-off-by: gmuselli <geoffrey.muselli@gmail.com>

@crenshaw-dev a small proposal for `patchTemplate`

Closes #11164
Closes https://github.com/argoproj/argo-cd/issues/9177

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by crenshaw-dev** [comment]
For folks watching this issue: I really, really wanted to get it into 2.9 but ran out of time. We did manage to merge this slightly less-powerful tool, which should be sufficient for some use cases: https://github.com/argoproj/argo-cd/pull/14743

**Comment by speedfl** [comment]
@crenshaw-dev #14743 will save a lot of time. Thanks a lot. In parallel I will rebase this one and keep it ready for 2.10

File: docs/operator-manual/applicationset/Template.md [hunk 1]
```diff
@@ -108,3 +108,60 @@ spec:
...
+
+Templating is only available on string type. However some uses cases may require to apply templating on other types.
+
+Example:
+
+- Set the automated sync policy
+- Switch prune boolean to true
+- Add multiple helm value files
+
+ArgoCD has a `patchTemplate` feature to allow advanced templating. It supports both json and yaml
```
Comment (sethgupton-mastery) [inline docs/operator-manual/applicationset/Template.md:122]:
```suggestion
ArgoCD has a `templatePatch` feature to allow advanced templating. It supports both json and yaml
```

I beat my head against the wall for a surprisingly long time trying to figure out what I was doing wrong. 🤣

File: docs/operator-manual/applicationset/Template.md [hunk 2]
```diff
@@ -108,3 +108,60 @@ spec:
...
+  - list:
+      elements:
+        - cluster: engineering-dev
+          url: https://kubernetes.default.svc
+          autoSync: true
+          prune: true
+          valueFiles:
+            - values.large.yaml
+            - values.debug.yaml
+  patchTemplate: |
```
Comment (sethgupton-mastery) [inline docs/operator-manual/applicationset/Template.md:142]:
```suggestion
  templatePatch: |
```

**Review by sethgupton-mastery — CHANGES_REQUESTED** [review]
Even with patchTemplate -> templatePatch I'm still having problems getting it working with my test files. But that could be on my end. Will continue on it next week.

File: docs/operator-manual/applicationset/Template.md (hunk 1 above)
Comment (speedfl) [reply docs/operator-manual/applicationset/Template.md:122]: Sorry about that. I'll correct it 😅

File: docs/operator-manual/applicationset/Template.md (hunk 2 above)
Comment (speedfl) [reply docs/operator-manual/applicationset/Template.md]: Done

File: docs/operator-manual/applicationset/Template.md (hunk 1 above)
Comment (speedfl) [reply docs/operator-manual/applicationset/Template.md]: Done

**Comment by speedfl** [comment]
@sethgupton-mastery as you are using goTemplate you must use `testfiles-argocd/**/{{ .folderName }}/app.json`

**Comment by sethgupton-mastery** [comment]
Sweet. That fixed me up. I hadn't realized goTemplate was something that already existed and wasn't part of this PR.
With goTemplate set to true and me using goTemplate correctly... *drum roll*

I was able to load a dynamic list of labels from my app.json file! Very exciting! Can't wait for this to get in.

Thanks speedfl.

**Review by crenshaw-dev — APPROVED** [review]
Let's go 🚀

**Comment by didlawowo** [comment]
i'm trying this feature using this article : https://medium.com/@geoffrey.muselli/argocd-multi-cluster-helm-charts-installation-in-mono-repo-0a406ff7c578

building an image based on 2.10, get for each time get Error from server (BadRequest): error when creating "demo/argo-applicationset-template/applicationset-argo-monocluster.yaml": ApplicationSet in version "v1alpha1" cannot be handled as a ApplicationSet: strict decoding error: unknown field "Generator"

**Comment by speedfl** [comment]
Can you open a bug ?
Please add as well the ApplicationSet you are trying to create and the logs from the appset controller ? You can tag me. I will try to have a look

**Comment by didlawowo** [comment]
> Can you open a bug ? Please add as well the ApplicationSet you are trying to create and the logs from the appset controller ? You can tag me. I will try to have a look

made a mistake in my config , sorry

**Comment by didlawowo** [comment]
it's working for multi cluster, i just make some little adjustement , but enable to use the feature flag enabled.

this is my applicationset, with that, i can get two clone on different cluster

`apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: hello-world-clusters
  namespace: kube-infra
spec:
  goTemplate: true
  generators:
  # Create a matrix generator that will install all chart in all clusters
  - matrix:
      generators:
      # Get all clusters
      - clusters: {}
        # selector:
        #   matchLabels:
        #     # Keep only the one which field enabled=true
        #     enabled: "true"
      - git:
          repoURL: git@github.com:didlawowo/pulumi.git
          revision: feat/crossplane-ml
          files:
          # Find all .argocd.json  in demo folder
          - path: 'continuous-delivery/clusters-apps/hello-world/.argocd.json'

  template:
    metadata:
      name: '{{  .name  }}-{{ .path.basename }}'
    spec:
      project: default
      sources:
      # First source is for value files
      # Default revision is HEAD but can be overriden with clusters.${name}.valuesRevision
      - repoURL: git@github.com:didlawowo/pulumi.git
        targetRevision: '{{ dig "clusters" .name "valuesRevision" "feat/crossplane-ml" . }}'
        ref: values
      # Second source is for chart
      # Default revision is .source.targetRevision but can be overriden with clusters.${name}.chartRevision
      - repoURL: '{{ .source.repoURL }}'
        targetRevision: '{{ dig "clusters" .name "chartRevision" .source.targetRevision . }}'
        chart: '{{ default "" .source.chart }}'
        path: '{{ default "" .source.path }}'
        helm:
          releaseName: '{{ .path.basename }}'
          valueFiles:
          - $values/{{ .path.path }}/values.yaml
          - $values/{{ .path.path }}/values.{{ .name }}.yaml
      destination:
        server: https://kubernetes.default.svc
        namespace: '{{ .destination.namespace }}'
      syncPolicy:
        automated:
          prune: true
        syncOptions:
        #- ServerSideApply={{ dig "syncPolicy" "syncOptions" "ServerSideApply" "true" . }}
        - CreateNamespace={{ dig "syncPolicy" "syncOptions" "CreateNamespace" "true" . }}

i have tried many tips, but not working, any help was really usefull

**Comment by speedfl** [comment]
This is a support question. Could you please ask it on the cncf slack. Section argo-cd-appset. Thanks 🙂

**Comment by LS80** [comment]
Thanks for implementing this.

Was any consideration given to the multiple source use case? i.e. the patching of a source that is a list element under `sources`? It would be great if it behaved like a Kustomize strategic merge patch and appended the patched source rather than replacing the entire contents of `sources`. It would be even better if it could merge the list elements on a known key (perhaps `ref`) so we would only have to patch one field. Then the bulk of the template could be under `template` as an object rather than a giant string under `templatePatch`.

An example of a desired ApplicationSet that doesn't currently work:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: vpa
  namespace: argocd
spec:
  goTemplate: true
  generators:
    - merge:
        mergeKeys:
          - name
        generators:
          - clusters: {}
          - list:
              elements:
              - name: dev
                kustomizeComponents: 
                  - components/dashboards
  templatePatch: |
    spec:
      sources:
        - ref: custom
          kustomize:
            components: 
            {{- range .kustomizeComponents }}
              - {{ . }}
            {{- end }}
  template:
    metadata:
      name: 'vpa-{{.name}}'
      namespace: argocd
      labels:
        environment: '{{.metadata.labels.environment}}'
    spec:
      sources:
        - repoURL: https://github.com/LS80/vertical-pod-autoscaler.git
          path: manifests/vpa
          targetRevision: '{{ default "main" .targetRevision }}'
          ref: custom
          kustomize:
            commonLabels:
              environment: '{{.metadata.labels.environment}}'
        - chart: vpa
          repoURL: https://charts.fairwinds.com/stable
          targetRevision: 2.1.0
          helm:
            releaseName: vpa
            valuesObject:
              updater:
                enabled: false
      project: default
      destination:
        name: '{{.name}}'
        namespace: kube-system
```

Here everything under `template.spec.sources` is lost. If it could work in a similar way to a Kustomize strategic merge patch for a Container in a Pod spec (uses the `name` as the merge key) then it would allow for a cleaner spec.

**Comment by speedfl** [comment]
This is a normal behavior but it could be probably improved.
The array in your `templatePatch` overrides the array in your sources.
The solution for array is to put everything in the `templatePatch`

**Comment by LS80** [comment]
Yes I am looking for a way to avoid putting everything in a giant string under `templatePatch`, similar to how Kustomize allows patching of a single field in a Container with a strategic merge patch, using the `name` as a merge key. How easy would it be to implement a similar merge key for `sources`?

**Comment by speedfl** [comment]
We are using a strategic merge patch so that strange

**Comment by speedfl** [comment]
Could you please open a bug. I'll try to have a look

**Comment by LS80** [comment]
I created https://github.com/argoproj/argo-cd/issues/16800. Thanks for looking 👍

**Pull request merged**